cpp/src/lists/copying/segmented_gather.cu

```diff
@@ -38,35 +41,46 @@ std::unique_ptr<column> segmented_gather(lists_column_view const& value_column,
   CUDF_EXPECTS(value_column.size() == gather_map.size(),
                "Gather map and list column should be same size");
 
-  auto gather_map_sliced_child = gather_map.get_sliced_child(stream);
-  auto const gather_map_size   = gather_map_sliced_child.size();
-  auto gather_index_begin      = gather_map.offsets().begin<size_type>() + 1 + gather_map.offset();
-  auto gather_index_end        = gather_index_begin + gather_map.size();
-  auto value_offsets           = value_column.offsets().begin<size_type>() + value_column.offset();
-  auto map_begin = cudf::detail::indexalator_factory::make_input_iterator(gather_map_sliced_child);
+  auto const gather_map_sliced_child = gather_map.get_sliced_child(stream);
+  auto const gather_map_size         = gather_map_sliced_child.size();
+  auto const gather_index_begin = gather_map.offsets().begin<size_type>() + 1 + gather_map.offset();
+  auto const gather_index_end   = gather_index_begin + gather_map.size();
+  auto const value_offsets      = value_column.offsets().begin<size_type>() + value_column.offset();
+  auto const map_begin =
+    cudf::detail::indexalator_factory::make_input_iterator(gather_map_sliced_child);
+  auto const out_of_bounds = [] __device__(auto const index, auto const list_size) {
+    return index >= list_size || (index < 0 && -index > list_size);
+  };
 
   // Calculate Flattened gather indices  (value_offset[row]+sub_index
-  auto transformer = [value_offsets, map_begin, gather_index_begin, gather_index_end] __device__(
-                       size_type index) -> size_type {
+  auto transformer = [value_offsets,
+                      map_begin,
+                      gather_index_begin,
+                      gather_index_end,
+                      bounds_policy,
+                      out_of_bounds] __device__(size_type index) -> size_type {
     // Get each row's offset. (Each row is a list).
     auto offset_idx =
       thrust::upper_bound(
         thrust::seq, gather_index_begin, gather_index_end, gather_index_begin[-1] + index) -
       gather_index_begin;
     // Get each sub_index in list in each row of gather_map.
-    auto sub_index         = map_begin[index];
-    auto list_size         = value_offsets[offset_idx + 1] - value_offsets[offset_idx];
-    auto wrapped_sub_index = (sub_index % list_size + list_size) % list_size;
+    auto sub_index          = map_begin[index];
+    auto list_size          = value_offsets[offset_idx + 1] - value_offsets[offset_idx];
+    auto wrapped_sub_index  = (sub_index % list_size + list_size) % list_size;
```

Comment:
I strongly recommend not using two modulos (they are very expensive).

```suggestion
    auto wrapped_sub_index  = sub_index < 0 ?  sub_index + list_size : sub_index; 
```

Reply:
Right you are. Sorry, I kept the preexisting phrasing.

Reply:
I have corrected this. But it brought another issue to light: 
The `segmented_gather()` [docs](https://github.com/rapidsai/cudf/blob/e245cd5feec0cc45b31d49e8173709aa116d4370/cpp/include/cudf/lists/gather.hpp#L46) indicate the following:
```c++
/* ...
 * If indices in `gather_map_list` are outside the range `[-n, n)`, where `n` is the number of
 * elements in corresponding row of the source column, the behavior is undefined.
 */
```
However, one sees the following in [`SegmentedGatherTest/GatherNested`](https://github.com/rapidsai/cudf/blob/e245cd5feec0cc45b31d49e8173709aa116d4370/cpp/tests/copying/segmented_gather_list_tests.cpp#L146):
```c++
    LCW<T> list{{{2, 3}, {4, 5}},
                {{6, 7, 8}, {9, 10, 11}, {12, 13, 14}},
                {{15, 16}, {17, 18}, {17, 18}, {17, 18}, {-17, -18}}};
    LCW<int> gather_map{{0, 2, -2}, {1}, {1, 0, -1, 5}};
```
Row#2 has only 5 entries. `segmented_gather()`'s output for index `5` should be undefined for row#2.

The previous modulo code was permissive, in that it returned element#0. I have now had to correct this. Could I please request for confirmation from @karthikeyann that there is no dependency on the previous behaviour?

Reply:
There is no dependency on this behavior. Modulo avoided Out of memory access. Above code will not avoid it. probably checking less than list_size might be useful.